crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,395 @@
+    const ESignerMismatch: u64 = 6;
+    /// Proposal (`KofNUpgradeCap`) and upgrade (`ProposedUpgrade`) do not match.
+    const EInvalidProposalForUpgrade: u64 = 7;
+
+    /// Create a `KofNUpgradeCap` given an `UpgradeCap`.
+    /// The returned instance is the only and exclusive controller of upgrades. 
+    /// The `k` (`required_votes`) out of `n` (length of `voters`) is set up
+    /// at construction time and it is immutable.
+    public fun new(
+        upgrade_cap: UpgradeCap,
+        required_votes: u64,
+        voters: VecSet<address>,
+        ctx: &mut TxContext,
+    ): KofNUpgradeCap {
+        // currently the allowed voters is limited to 100 and the number of
+        // required votes must be at least 2 and less or equal than the number of voters
+        assert!(vec_set::size(&voters) > 1, EAllowedVotersError);
+        assert!(vec_set::size(&voters) <= 100, EAllowedVotersError);
+        assert!(required_votes > 0, ERequiredVotesError);
+        assert!(required_votes <= vec_set::size(&voters), ERequiredVotesError);
+
+        // upgrade cap id
+        let cap_uid = object::new(ctx);
+        let cap_id = object::uid_to_inner(&cap_uid);
+
+        let voter_caps: VecSet<ID> = vec_set::empty();
+        let voter_addresses = vec_set::keys(&voters);
+        let voter_idx = vector::length(voter_addresses);
+        while (voter_idx > 0) {
+            voter_idx = voter_idx - 1;
+            let address = *vector::borrow(voter_addresses, voter_idx);
+            let voter_uid = object::new(ctx);
+            let voter_id = object::uid_to_inner(&voter_uid);
+            transfer::transfer(
+                VotingCap {
+                    id: voter_uid,
+                    owner: address,
+                    upgrade_cap: cap_id,
+                    transfers_count: 0,
+                    votes_issued: 0,
+                },
+                address,
+            );
+            vec_set::insert(&mut voter_caps, voter_id);
+        };
+
+        KofNUpgradeCap {
+            id: cap_uid,
+            upgrade_cap,
+            required_votes,
+            voters,
+            voter_caps,
+        }
+    }
+
+    /// Make the package immutable by destroying the k of n upgrade cap and the
+    /// underlying upgrade cap.
+    public fun make_immutable(cap: KofNUpgradeCap) {
+        let KofNUpgradeCap {
+            id,
+            upgrade_cap,
+            required_votes: _,
+            voters: _,
+            voter_caps: _,
+        } = cap;
+        object::delete(id);
+        package::make_immutable(upgrade_cap);
+    }
+
+    /// Restrict upgrades to "add code only", or "change dependencies".
+    public fun only_additive_upgrades(cap: &mut KofNUpgradeCap) {
+        package::only_additive_upgrades(&mut cap.upgrade_cap)
+    }
+
+    /// Restrict upgrades to "change dependencies only".
+    public fun only_dep_upgrades(cap: &mut KofNUpgradeCap) {
+        package::only_dep_upgrades(&mut cap.upgrade_cap)
+    }
```

Comment:
If we leave these APIs unguarded, then when we create the upgrade cap, we should transfer it directly to the sender, to prevent it being shared.

Reply:
maybe? I guess I do not understand how different it is from the regular publish path, and how much can we really protect with an object that has `key, store`.
Thoughts?

Reply:
done